src/js/components/TableCell/TableCell.js

```diff
@@ -125,49 +125,58 @@ const TableCell = forwardRef(
       );
     }
 
-    const themeProviderValue = { ...theme };
+    content = (
+      <StyledTableCell
+        ref={cellRef}
+        as={scope ? 'th' : undefined}
+        scope={scope}
+        size={size}
+        colSpan={colSpan}
+        tableContext={tableContext}
+        tableContextTheme={tableContextTheme}
+        {...(plain === true ? mergedProps : {})}
+        {...cellProps}
+        className={className}
+      >
+        {plain || !Object.keys(mergedProps).length ? (
+          content
+        ) : (
+          <Box
+            {...mergedProps}
+            align={align}
+            justify={verticalAlignToJustify[verticalAlign]}
+          >
+            {children}
+          </Box>
+        )}
+      </StyledTableCell>
+    );
 
     if (cellProps.background || theme.darkChanged) {
+      let themeProviderValue;
       const dark = backgroundIsDark(cellProps.background, theme);
       const darkChanged = dark !== undefined && dark !== theme.dark;
       if (darkChanged || theme.darkChanged) {
+        themeProviderValue = { ...theme };
         themeProviderValue.dark = dark === undefined ? theme.dark : dark;
         themeProviderValue.background = cellProps.background;
       } else if (cellProps.background) {
         // This allows DataTable to intelligently set the background of a pinned
         // header or footer.
+        themeProviderValue = { ...theme };
         themeProviderValue.background = cellProps.background;
       }
+
+      if (themeProviderValue) {
```

Comment:
I don't think you should do this conditionally or we could easily run into something like #5146. If the color changes or light vs dark theme etc, you could cause a new hierarchy which causes the components within this cell to get re-created, losing any state or focus, etc.

Reply:
good catch, reverted